File: Difficulty: Medium/0 - 1 Knapsack Problem/0-1-knapsack-problem.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
    vector<vector<int>> dp;
    int fun(int idx, vector<int> &val, vector<int> &wt, int &cap){
        if(idx >= val.size()) 
            return 0;
        else if(dp[idx][cap] != -1)
            return dp[idx][cap];
        
        int take = 0, notTake = fun(idx+1, val, wt, cap);
        if(cap >= wt[idx]){
            cap -= wt[idx];
            take = fun(idx+1, val, wt, cap) + val[idx];
            cap += wt[idx];
        }
        return dp[idx][cap] = max(take, notTake);
    }
  public:
    // Function to return max value that can be put in knapsack of capacity.
    int knapsack(int W, vector<int> &val, vector<int> &wt) {
        // code here
        dp.resize(val.size(), vector<int>(W+1, -1));
        return fun(0, val, wt, W);
    }
};


//{ Driver Code Starts.

int main() {
    // Taking total test cases
    int testCases;
    cin >> testCases;
    cin.ignore();
    while (testCases--) {
        // Reading number of items and capacity
        int numberOfItems, capacity;
        vector<int> weights, values;
        string input;
        int number;

        // Read capacity and number of items
        getline(cin, input);
        stringstream ss(input);
        ss >> capacity;      // The first number is the capacity
        ss >> numberOfItems; // The second number is the number of items

        // Read values
        getline(cin, input);
        ss.clear();
        ss.str(input);
        while (ss >> number) {
            values.push_back(number);
        }

        // Read weights
        getline(cin, input);
        ss.clear();
        ss.str(input);
        while (ss >> number) {
            weights.push_back(number);
        }

        Solution solution;
        cout << solution.knapsack(capacity, values, weights) << endl;
        cout << "~" << endl;
    }
    return 0;
}

// } Driver Code Ends

File: Difficulty: Medium/0 - 1 Knapsack Problem/0-1-knapsack-problem.py
#User function Template for python3

class Solution:
    
    #Function to return max value that can be put in knapsack of capacity W.
    def knapSack(self,W, wt, val):
        dp = [[-1]*(W+1) for _ in range(len(wt))]
        def fun(i, cap):
            nonlocal wt
            nonlocal val
            nonlocal dp
            if i == len(wt): return 0
            if dp[i][cap] != -1: return dp[i][cap]
            
            notTake = fun(i+1, cap)
            take = (fun(i+1, cap - wt[i]) + val[i]) if cap >= wt[i] else 0
            
            dp[i][cap] = max(take, notTake)
            return dp[i][cap]
            
        return fun(0, W)


#{ 
 # Driver Code Starts
#Initial Template for Python 3
import atexit
import io
import sys

if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases):
        # n = int(input())
        W = int(input())
        val = list(map(int, input().strip().split()))
        wt = list(map(int, input().strip().split()))
        ob = Solution()
        print(ob.knapSack(W, wt, val))

# } Driver Code Ends